### ISSUE TITLE
SQL syntax error while filtering participants

### ISSUE THREAD

**Issue by mrkskwsnck** [academic-moodle-cooperation/moodle-mod_publication#63]
Moodle version: 3.11 [branch: MOODLE_311_STABLE]
Plug-in version: 3.11.3 [branch: MOODLE_311_STABLE]
DBMS: Microsoft SQL Server [driver: sqlsrv]

**Error message**

```
Error reading from database

More information about this error
Debug info: SQLState: 42000<br>
Error Code: 156<br>
Message: [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Incorrect syntax near the keyword 'AND'.<br>

SELECT COUNT(a.uid) FROM (SELECT DISTINCT u.id AS uid FROM mdl_user u LEFT JOIN mdl_publication_file files ON u.id = files.userid AND files.publication = '23619' WHERE u.id IN ('7650','7707','8441','8511','9136','16351','17122','17339','18408','19390','24318','29681','32866','33688','34576','35893','37466','40221','40957','41184','44317','47958','49820','50172','51032','53059','55367','60232','60404','60443','60665','60983','61097','61880','62181','62366','62535','62617','62963','62973','63119','63294','64248','64369','64411','64531','64793','64941','65134','65406','66212','66240','66316','621485','640259','640261','654548','655078','657380','657386','657390','657406','657534','657614','657628','659806','659891','660135','660171','660200','660778','662780','704270','708770','709051','709076','710261','710465','710504','710512','710934','711000','711758','712314','714680','714788','714790','1042361')) a AND firstname COLLATE Latin1_General_CI_AI LIKE N'E%' ESCAPE '\'
[array (
0 => '23619',
1 => '7650',
2 => '7707',
3 => '8441',
4 => '8511',
5 => '9136',
6 => '16351',
7 => '17122',
8 => '17339',
9 => '18408',
10 => '19390',
11 => '24318',
12 => '29681',
13 => '32866',
14 => '33688',
15 => '34576',
16 => '35893',
17 => '37466',
18 => '40221',
19 => '40957',
20 => '41184',
21 => '44317',
22 => '47958',
23 => '49820',
24 => '50172',
25 => '51032',
26 => '53059',
27 => '55367',
28 => '60232',
29 => '60404',
30 => '60443',
31 => '60665',
32 => '60983',
33 => '61097',
34 => '61880',
35 => '62181',
36 => '62366',
37 => '62535',
38 => '62617',
39 => '62963',
40 => '62973',
41 => '63119',
42 => '63294',
43 => '64248',
44 => '64369',
45 => '64411',
46 => '64531',
47 => '64793',
48 => '64941',
49 => '65134',
50 => '65406',
51 => '66212',
52 => '66240',
53 => '66316',
54 => '621485',
55 => '640259',
56 => '640261',
57 => '654548',
58 => '655078',
59 => '657380',
60 => '657386',
61 => '657390',
62 => '657406',
63 => '657534',
64 => '657614',
65 => '657628',
66 => '659806',
67 => '659891',
68 => '660135',
69 => '660171',
70 => '660200',
71 => '660778',
72 => '662780',
73 => '704270',
74 => '708770',
75 => '709051',
76 => '709076',
77 => '710261',
78 => '710465',
79 => '710504',
80 => '710512',
81 => '710934',
82 => '711000',
83 => '711758',
84 => '712314',
85 => '714680',
86 => '714788',
87 => '714790',
88 => '1042361',
89 => 'E%',
)]
Error code: dmlreadexception
Stack trace:

    line 486 of /lib/dml/moodle_database.php: dml_read_exception thrown
    line 324 of /lib/dml/sqlsrv_native_moodle_database.php: call to moodle_database->query_end()
    line 431 of /lib/dml/sqlsrv_native_moodle_database.php: call to sqlsrv_native_moodle_database->query_end()
    line 900 of /lib/dml/sqlsrv_native_moodle_database.php: call to sqlsrv_native_moodle_database->do_query()
    line 976 of /lib/dml/sqlsrv_native_moodle_database.php: call to sqlsrv_native_moodle_database->get_recordset_sql()
    line 1671 of /lib/dml/moodle_database.php: call to sqlsrv_native_moodle_database->get_records_sql()
    line 1744 of /lib/dml/moodle_database.php: call to moodle_database->get_record_sql()
    line 1954 of /lib/dml/moodle_database.php: call to moodle_database->get_field_sql()
    line 285 of /mod/publication/classes/local/allfilestable/base.php: call to moodle_database->count_records_sql()
    line 2091 of /lib/tablelib.php: call to mod_publication\local\allfilestable\base->query_db()
    line 498 of /mod/publication/locallib.php: call to table_sql->out()
    line 288 of /mod/publication/view.php: call to publication->display_allfilesform()
```

**Description of problem**

¿It seems the extension to the where clause at https://github.com/academic-moodle-cooperation/moodle-mod_publication/blob/78af484507bc4fa40164e0a1244b0b1a446ff820/classes/local/allfilestable/base.php#L279 is appended to the outer instead of the inner SELECT statement of `countsql`? However, the root cause might be located at https://github.com/academic-moodle-cooperation/moodle-mod_publication/blob/78af484507bc4fa40164e0a1244b0b1a446ff820/classes/local/allfilestable/base.php#L237 because the SQL statement is a little bit too static. IMHO

**Hack/Patch**

So far, I did not develop a PR worthy patch. Instead, I provide a hack that _works for me_ :-)

```diff
diff --git a/classes/local/allfilestable/base.php b/classes/local/allfilestable/base.php
index 9dc47d5..9c8c608 100644
--- a/classes/local/allfilestable/base.php
+++ b/classes/local/allfilestable/base.php
@@ -276,7 +276,7 @@ class base extends \table_sql {
 
             list($wsql, $wparams) = $this->get_sql_where();
             if ($wsql) {
-                $this->countsql .= ' AND ' . $wsql;
+                $this->countsql = preg_replace('/\)\)/', ") AND {$wsql})", $this->countsql);    # HACK Appending to inner instead of outer statement
                 $this->countparams = array_merge($this->countparams, $wparams);
 
                 $this->sql->where .= ' AND ' . $wsql;
```

**Comment by goggo24** [comment]
The error occurs when the table is filtered by first and/or last name. It will be fixed with v3.11.4 - coming in the next days. 
---
Workaround:   the table can be reset if the addition &treset=1 is entered in the browser address line (if you are in the student folder), e.g. https://moodle_xyz/mod/publication/view.php?id=12345&treset= 1

**Comment by goggo24** [comment]
Just published - have fun!
